jobs/kpi-forecasting/kpi_forecasting/models/funnel_forecast.py

```diff
@@ -17,13 +32,590 @@ class SegmentModelHolder:
+            on="submission_date",
+            how="left",
+        ).fillna(0)
+
+        forecast_summarized = (
+            forecast_agg.set_index("submission_date")
+            # Add observed data samples to any overlapping forecasted period. This
+            # ensures that any forecast made partway through a period accounts for
+            # previously observed data within the period. For example, when a monthly
+            # forecast is generated in the middle of the month.
+            .add(overlap[["value"]].values)
+            # calculate summary values, aggregating by submission_date,
+            .agg(aggregations, axis=1).reset_index()
+        ).rename(columns=self._percentile_name_map(percentiles))
+
+        # add datasource-specific metadata columns
+        forecast_summarized["source"] = "forecast"
+        observed_summarized["source"] = "historical"
+
+        # add segment columns to forecast  table
+        for dim, value in segment_settings.segment.items():
+            forecast_summarized[dim] = value
+
+        # create a single dataframe that contains observed and forecasted data
+        df = pd.concat([observed_summarized, forecast_summarized])
+
+        # add summary metadata columns
+        df["aggregation_period"] = period.lower()
+
+        # reorder columns to make interpretation easier
+        df = df[
+            [
+                "submission_date",
+                "aggregation_period",
+                "source",
+                "value",
+                "value_low",
+                "value_mid",
+                "value_high",
+            ]
+        ]
+
+        # add Metric Hub metadata columns
+        df["metric_alias"] = self.metric_hub.alias.lower()
+        df["metric_hub_app_name"] = self.metric_hub.app_name.lower()
+        df["metric_hub_slug"] = self.metric_hub.slug.lower()
+        df["metric_start_date"] = pd.to_datetime(self.metric_hub.min_date)
+        df["metric_end_date"] = pd.to_datetime(self.metric_hub.max_date)
+        df["metric_collected_at"] = self.collected_at
+
+        # add forecast model metadata columns
+        df["forecast_start_date"] = self.start_date
+        df["forecast_end_date"] = self.end_date
+        df["forecast_trained_at"] = self.trained_at
+        df["forecast_predicted_at"] = self.predicted_at
+        df["forecast_parameters"] = json.dumps(segment_settings.trained_parameters)
+
+        return df
+
+    def predict(self) -> None:
+        """Generate a forecast from `start_date` to `end_date`."""
+        print(f"Forecasting from {self.start_date} to {self.end_date}.", flush=True)
+        self._set_seed()
+        self.predicted_at = datetime.utcnow()
+
+        for segment_settings in self.segment_models:
+            forecast_df = self._predict(segment_settings)
+            self._validate_forecast_df(forecast_df)
+
+            segment_settings.forecast_df = forecast_df.copy(deep=True)
+            del forecast_df
```

Comment:
Why do you need to copy and manually delete here?

Reply:
Hmm, probably my paranoia. I'll remove it